# Group 4 - Final Report

Dữ liệu BankX Marketing

## Import libraries

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, precision_recall_curve
import matplotlib.pyplot as plt
import seaborn as sns

## 1. Loading data
**Read in the bankX.csv file and set it to a data frame called df**

In [3]:
df = pd.read_csv('bankX.csv', header = 0)

**Explore data**

In [3]:
df.head()

Unnamed: 0  age           job  marital  education default  balance housing  \
0           0   58    management  married   tertiary      no     2143     yes   
1           1   44    technician   single  secondary      no       29     yes   
2           2   33  entrepreneur  married  secondary      no        2     yes   
3           3   47   blue-collar  married    unknown      no     1506     yes   
4           4   33       unknown   single    unknown      no        1      no   

  loan  contact  day month  duration  campaign  pdays  previous poutcome   y  
0   no  unknown    5   may       261         1     -1         0  unknown  no  
1   no  unknown    5   may       151         1     -1         0  unknown  no  
2  yes  unknown    5   may        76         1     -1         0  unknown  no  
3   no  unknown    5   may        92         1     -1         0  unknown  no  
4   no  unknown    5   may       198         1     -1         0  unknown  no

**Delete redundant columns**

In [4]:
df.drop(['Unnamed: 0'], axis=1, inplace=True)
df.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no

**Delete the redundant dots in 'admin.' values of job column**

In [5]:
df['job'].replace("admin.", "admin", inplace = True)
df.head()

C:\Users\User\AppData\Local\Temp\ipykernel_6108\1249272988.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['job'].replace("admin.", "admin", inplace = True)


age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no

## 2. Data Wrangling

### 2.1. Identify and handle missing values

**Identify missing values**

In the bankX.csv dataset, missing data comes with the word "unknown". We replace "unknown" with NaN (Not a Number), which is Pandas's default missing value marker, for reasons of computational speed and convenience. 

In [6]:
df.replace("unknown", np.nan, inplace = True)
df.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married        NaN      no     1506     yes   no   
4   33           NaN   single        NaN      no        1      no   no   

  contact  day month  duration  campaign  pdays  previous poutcome   y  
0     NaN    5   may       261         1     -1         0      NaN  no  
1     NaN    5   may       151         1     -1         0      NaN  no  
2     NaN    5   may        76         1     -1         0      NaN  no  
3     NaN    5   may        92         1     -1         0      NaN  no  
4     NaN    5   may       198         1     -1         0      NaN  no

**Count of missing values in each column**

In [7]:
df.isnull().sum(axis=0)

age              0
job            288
marital          0
education     1857
default          0
balance          0
housing          0
loan             0
contact      13020
day              0
month            0
duration         0
campaign         0
pdays            0
previous         0
poutcome     36959
y                0
dtype: int64

Based on the summary above, four columns containing missing data:
- "job": 288 missing values
- "education": 1857 missing values
- "contact": 13020 missing values
- "poutcome": 36959 missing values

**Deal with missing data**

Because the number of missing values in the output column is too large (greater than 50%), we decided to remove this column in the model.

In [9]:
df.drop("poutcome", axis = 1, inplace=True)
df.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married        NaN      no     1506     yes   no   
4   33           NaN   single        NaN      no        1      no   no   

  contact  day month  duration  campaign  pdays  previous   y  
0     NaN    5   may       261         1     -1         0  no  
1     NaN    5   may       151         1     -1         0  no  
2     NaN    5   may        76         1     -1         0  no  
3     NaN    5   may        92         1     -1         0  no  
4     NaN    5   may       198         1     -1         0  no

The missing data rate for the two columns job and education was not too large, so the team decided to replace it with the mode value.

In [10]:
df['job'].value_counts()

job
blue-collar      9732
management       9458
technician       7597
admin            5171
services         4154
retired          2264
self-employed    1579
entrepreneur     1487
unemployed       1303
housemaid        1240
student           938
Name: count, dtype: int64

In [11]:
df['education'].value_counts()

education
secondary    23202
tertiary     13301
primary       6851
Name: count, dtype: int64

In [8]:
df['job'].fillna('blue-collar', inplace=True)
df['education'].fillna('secondary', inplace=True)
df.head()

C:\Users\User\AppData\Local\Temp\ipykernel_6108\2859828306.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['job'].fillna('blue-collar', inplace=True)
C:\Users\User\AppData\Local\Temp\ipykernel_6108\2859828306.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, w

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married  secondary      no     1506     yes   no   
4   33   blue-collar   single  secondary      no        1      no   no   

  contact  day month  duration  campaign  pdays  previous poutcome   y  
0     NaN    5   may       261         1     -1         0      NaN  no  
1     NaN    5   may       151         1     -1         0      NaN  no  
2     NaN    5   may        76         1     -1         0      NaN  no  
3     NaN    5   may        92         1     -1         0      NaN  no  
4     NaN    5   may       198         1     -1         0      NaN  no

In the contact column, we see that the 'cellular' is dominant. So we decided to fill it in missing values

In [13]:
df['contact'].value_counts()

contact
cellular     29285
telephone     2906
Name: count, dtype: int64

In [14]:
df['contact'].fillna('cellular', inplace=True)
df.head()

C:\Users\User\AppData\Local\Temp\ipykernel_13780\1162142702.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['contact'].fillna('cellular', inplace=True)


age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married  secondary      no     1506     yes   no   
4   33   blue-collar   single  secondary      no        1      no   no   

    contact  day month  duration  campaign  pdays  previous   y  
0  cellular    5   may       261         1     -1         0  no  
1  cellular    5   may       151         1     -1         0  no  
2  cellular    5   may        76         1     -1         0  no  
3  cellular    5   may        92         1     -1         0  no  
4  cellular    5   may       198         1     -1         0  no

### 2.2 Correct data format

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        45211 non-null  int64 
 1   job        45211 non-null  object
 2   marital    45211 non-null  object
 3   education  45211 non-null  object
 4   default    45211 non-null  object
 5   balance    45211 non-null  int64 
 6   housing    45211 non-null  object
 7   loan       45211 non-null  object
 8   contact    32191 non-null  object
 9   day        45211 non-null  int64 
 10  month      45211 non-null  object
 11  duration   45211 non-null  int64 
 12  campaign   45211 non-null  int64 
 13  pdays      45211 non-null  int64 
 14  previous   45211 non-null  int64 
 15  poutcome   8252 non-null   object
 16  y          45211 non-null  object
dtypes: int64(7), object(10)
memory usage: 5.9+ MB


We see that the features are of the correct data type.

Now, we finally obtain the cleaned dataset with no missing values and all data in its proper format.

### 2.3 Dummy Variable

Binary column encoding: replace "yes" and "no" in columns to 1 and 0

In [16]:
df['subcribe'] = np.where(df['y'] == 'yes', 1, 0)
df.drop(['y'], axis=1, inplace=True)

In [17]:
df['default1'] = np.where(df['default'] == 'yes', 1, 0)
df.drop(['default'], axis=1, inplace=True)

In [18]:
df['loan1'] = np.where(df['loan'] == 'yes', 1, 0)
df.drop(['loan'], axis=1, inplace=True)

In [19]:
df['housing1'] = np.where(df['housing'] == 'yes', 1, 0)
df.drop(['housing'], axis=1, inplace=True)

In [10]:
df.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married  secondary      no     1506     yes   no   
4   33   blue-collar   single  secondary      no        1      no   no   

  contact  day month  duration  campaign  pdays  previous poutcome   y  
0     NaN    5   may       261         1     -1         0      NaN  no  
1     NaN    5   may       151         1     -1         0      NaN  no  
2     NaN    5   may        76         1     -1         0      NaN  no  
3     NaN    5   may        92         1     -1         0      NaN  no  
4     NaN    5   may       198         1     -1         0      NaN  no

Change the month column value from strings to intergers.

ValueError: time data "jun" doesn't match format "%B", at position 1. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

In [11]:
## biến education thì getdummy() 3 biến nhỏ
## biến job: thì retired, self-employed, unemployed và housemaid --> 0 (ý là kiểu người ta đang ko làm việc trong 1 cơ quan), còn lại thì sẽ là 1
df['marital'].value_counts()
## marital thì married là 1, còn single divorced là 0

marital
married     27214
single      12790
divorced     5207
Name: count, dtype: int64

In [12]:
## biến education thì getdummy() 3 biến nhỏ
education_dummies = pd.get_dummies(df['education'], prefix='education')

df = pd.concat([df, education_dummies], axis=1)
df.drop('education', axis=1, inplace=True)

print(df.head())

   age           job  marital default  balance housing loan contact  day  \
0   58    management  married      no     2143     yes   no     NaN    5   
1   44    technician   single      no       29     yes   no     NaN    5   
2   33  entrepreneur  married      no        2     yes  yes     NaN    5   
3   47   blue-collar  married      no     1506     yes   no     NaN    5   
4   33   blue-collar   single      no        1      no   no     NaN    5   

  month  duration  campaign  pdays  previous poutcome   y  education_primary  \
0   may       261         1     -1         0      NaN  no              False   
1   may       151         1     -1         0      NaN  no              False   
2   may        76         1     -1         0      NaN  no              False   
3   may        92         1     -1         0      NaN  no              False   
4   may       198         1     -1         0      NaN  no              False   

   education_secondary  education_tertiary  
0                

In [ ]:
## biến job: thì retired, self-employed, unemployed và housemaid --> 0 (ý là kiểu người ta đang ko làm việc trong 1 cơ quan), còn lại thì sẽ là
# Tạo cột mới 'in_company' với giá trị 1 hoặc 0 dựa trên điều kiện
non_company_jobs = ['retired', 'self-employed', 'unemployed', 'housemaid']
df['in_company'] = df['job'].apply(lambda x: 0 if x in non_company_jobs else 1) 
# print test
print(df[['job', 'in_company']])


                job  in_company
0        management           1
1        technician           1
2      entrepreneur           1
3       blue-collar           1
4       blue-collar           1
...             ...         ...
45206    technician           1
45207       retired           0
45208       retired           0
45209   blue-collar           1
45210  entrepreneur           1

[45211 rows x 2 columns]


In [ ]:
# marital thì married là 1, còn single divorced là 0
# Tạo cột mới 'is_married' với giá trị 1 hoặc 0 dựa trên điều kiện
df['is_married'] = df['marital'].apply(lambda x: 1 if x == 'married' else 0)
df.drop(columns=['marital'], inplace=True)

print(df.head())

   age           job default  balance housing loan contact  day month  \
0   58    management      no     2143     yes   no     NaN    5   may   
1   44    technician      no       29     yes   no     NaN    5   may   
2   33  entrepreneur      no        2     yes  yes     NaN    5   may   
3   47   blue-collar      no     1506     yes   no     NaN    5   may   
4   33   blue-collar      no        1      no   no     NaN    5   may   

   duration  campaign  pdays  previous poutcome   y  education_primary  \
0       261         1     -1         0      NaN  no              False   
1       151         1     -1         0      NaN  no              False   
2        76         1     -1         0      NaN  no              False   
3        92         1     -1         0      NaN  no              False   
4       198         1     -1         0      NaN  no              False   

   education_secondary  education_tertiary  in_company  is_married  
0                False                True     